man/repart.d.xml

```diff
@@ -551,6 +551,29 @@
         factory reset operation. This functionality is useful to implement schemes where images can be reset
         into their original state by removing partitions and creating them anew. Defaults to off.</para></listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><varname>Flags=</varname></term>
+
+        <listitem><para>Configures the 64bit GPT partition flags to set for the partition when creating
+        it. This option has no effect if the partition already exists. If not specified the flags values is
+        set to all zeroes, except if the partition type (as configured with <varname>Type=</varname> above)
+        refers to a Verity partition, in wich case bit 60 is set (i.e. the read-only bit). This bit may also
+        be configured separately via <varname>ReadOnly=</varname>, see below. Specify the flags value in
+        hexadecimal (by prefixing it with <literal>0x</literal>), binary (prefix <literal>0b</literal>) or
+        decimal (no prefix).</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>ReadOnly=</varname></term>
+
+        <listitem><para>Configures the Read-Only partition flags (bit 60) of the partition table entry. This
+        option is a friendly way to set bit 60 of the partition flags value without setting any of the other
+        bits, and may be set via <varname>Flags=</varname> too, see above.</para>
+
+        <para>If both <varname>Flags=</varname> and <varname>ReadOnly=</varname> are set the latter controls
+        the value of the flag.</para></listitem>
```

Comment:
Hmm, does this redundancy make sense? `Flags=read-only` would seem just fine... And then we wouldn't have to explain which setting takes precedence, etc. And even if flags are inherited, this flag can be treated appropriately, e.g. `Flags=inherit-from-source read-only`.

Reply:
So the reason why I made this two distinct things is because one is more high-level than the other. i.e. the way I figured this should be used is so that Flags= is a low-level bit-by-bit concept: you specify some bits of your choice there, we won't second guess them, and you can define whatever you want there regardless if it makes sense is defined in any spec or otherwise. Most of the time people shouldn't bother.

The ReadOnly= flag otoh is supposed to be higher level, it's what you actually want to use. The idea is that it tries to make sense of the situation and will set the right flag for you in a safer way.

This reflects the fact that the R/O flag is a private per partition type flag, i.e. the types we defined ourselves know it, and Microsoft Windows generic partitions have a flag for it too, but not much else. ESP doesn't (and has another flag at its place), Chromebook partitions don't either (and put a boot counter in its place). Thus, ReadOnly= is supposed to come with some extra intelligence, so that it maps to the right flag on the right partition types and if the partition type doesn't have it, doesn't do it.

Now, right now we pretty much support only our own partition types, but the idea is that this eventually opened up maybe. And for those the mapping for ReadOnly= to specific flag might differ or be masked away, too.

i.e. internally we have to keep the flag separate anyway, since we don't know which bit to actually set until Type= is parsed too. Thus in order to be able to parse the r/o flag and Type= in either order we need to store the r/o flag separately, and only OR it into the flags value after we parsed the whole file.

Thus: ReadOnly= is what I myself would typically use when I want to build a read-only image. But Flags= is the option that people who want to build cross images that also contain an Android, Windows, FreeBSD image or similar, which have very different flags definitions.

Hope that makes sense?

Reply:
OK.